src/v/cluster/types.h

```diff
@@ -2979,6 +2979,41 @@ struct cancel_partition_movements_reply
     std::vector<move_cancellation_result> partition_results;
 };
 
+struct cloud_storage_usage_request
+  : serde::envelope<
+      cloud_storage_usage_request,
+      serde::version<0>,
+      serde::compat_version<0>> {
+    using rpc_adl_exempt = std::true_type;
+
+    absl::flat_hash_map<ss::shard_id, std::vector<model::ntp>> partitions;
```

Comment:
In the current form, I think it does. If we don't include the expected shard assignment, we would have to iterate through the entire topic table on each broker which is potentially expensive with 100ks of partitions.

You might ask: "Why do we need to iterate over the topic table in that case? We could just reduce over the partition manager as we do now." The issue here is that the topic manager lags behind the state in the topic table. When a reconfiguration is made, the topic table is updated via `replicate_and_wait` and generates deltas for the controller backend. These deltas are handled async and result in updates to the partition manager. If we relied on the contents of the partition manager, we can miss usage for partitions that have just been moved (i.e. state was updated in the topic table, but the ntp reconcilication in the controller backend did not catch up yet).

TLDR: It's a trade-off. We either iterate over all the ntps when creating the request (on one node) or when handling the request (on all nodes).

Reply:
Sorry I meant `shard_table` not `topic_table`. The `shard_table` is a map, lookups for `shard_id` are quick. So i'm suggesting we do send the `model::ntp` to the peer so all of what you've mentioned continues to work, with the exception of also sending the `shard_id` with the `ntp`

Reply:
I think going via the `shard_table` works because it's updated together with the `partition_manager` in the controller backend reconciliation. Will give it a go.

Reply:
Updated the PR.